assets/js/components/dashboard-sharing/ModuleRecoveryAlert.js

```diff
@@ -0,0 +1,230 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, useCallback, useEffect, useState } from '@wordpress/element';
+import { sprintf, __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_MODULES } from '../../googlesitekit/modules/datastore/constants';
+import Button from '../Button';
+import Checkbox from '../Checkbox';
+import BannerNotification from '../notifications/BannerNotification';
+import ProgressBar from '../ProgressBar';
+import Spinner from '../Spinner';
+
+const { useDispatch, useSelect } = Data;
+
+export default function ModuleRecoveryAlert() {
+	const [ checkboxes, setCheckboxes ] = useState( null );
+	const [ recoveringModules, setRecoveringModules ] = useState( false );
+
+	const recoverableModules = useSelect( ( select ) =>
+		select( CORE_MODULES ).getRecoverableModules()
+	);
+
+	const userAccessibleModules = useSelect( ( select ) => {
+		const modules = select( CORE_MODULES ).getRecoverableModules();
+
+		if ( modules === undefined ) {
+			return undefined;
+		}
+
+		const accessibleModules = Object.keys( modules ).map( ( slug ) => ( {
+			slug,
+			hasModuleAccess: select( CORE_MODULES ).hasModuleAccess( slug ),
+		} ) );
+
+		if (
+			accessibleModules.some(
+				( { hasModuleAccess } ) => hasModuleAccess === undefined
+			)
+		) {
+			return undefined;
+		}
+
+		return accessibleModules
+			.filter( ( { hasModuleAccess } ) => hasModuleAccess )
+			.map( ( { slug } ) => slug );
+	} );
+
+	const { recoverModule } = useDispatch( CORE_MODULES );
+
+	const updateCheckboxes = useCallback(
+		( slug ) =>
+			setCheckboxes( ( currentCheckboxes ) => ( {
+				...currentCheckboxes,
+				[ slug ]: ! currentCheckboxes[ slug ],
+			} ) ),
+		[]
+	);
+
+	const handleRecoverModules = useCallback( () => {
+		setRecoveringModules( true );
+		const modulesToRecover = Object.keys( checkboxes ).filter(
+			( module ) => checkboxes[ module ]
+		);
+		Promise.all(
+			modulesToRecover.map( ( module ) => recoverModule( module ) )
+		).finally( () => setRecoveringModules( false ) );
+	}, [ checkboxes, recoverModule ] );
+
+	useEffect( () => {
+		if ( userAccessibleModules !== undefined && checkboxes === null ) {
+			const checked = {};
+
+			userAccessibleModules.forEach( ( module ) => {
+				checked[ module ] = true;
+			} );
+
+			setCheckboxes( checked );
+		}
+	}, [ checkboxes, userAccessibleModules ] );
+
+	const recoverableModulesList = Object.keys( recoverableModules );
+
+	if (
+		recoverableModules === undefined ||
+		recoverableModulesList.length === 0
+	) {
+		return null;
+	}
```

Comment:
If we know that `recoverableModules` may be undefined, then we should check it before trying to get keys from it because otherwise it may cause a TypeError exception.

```suggestion
	const recoverableModulesList = Object.keys( recoverableModules || {} );

	if (
		recoverableModules === undefined ||
		recoverableModulesList.length === 0
	) {
		return null;
	}
```